### PR TITLE
Add multiqc

### PR DESCRIPTION
**Pull request by rroutsong**
Add multiqc into ngsqc pipeline.

Addresses #12 #7

### CODE REVIEW

**Comment by rroutsong** [comment]
@skchronicles , example report at `/data/RTB_GRS/dev/Dmux/test_ngsqc2/GRS_0212_Bhasym/230907_NS500353_0215_AHLTNVBGXM/multiqc/Run-230907_NS500353_0215_AHLTNVBGXM-Project-GRS_0212_Bhasym_multiqc_report.html`

File: src/Dmux/workflow/ngs_qaqc/fastq.smk [hunk 1]
```diff
@@ -42,7 +41,6 @@ rule fastq_screen:
         subset              = 1000000,
         aligner             = "bowtie2",
         output_dir          = lambda w: config['out_to'] + "/" + w.project + "/" + config['run_ids'] + "/" + w.sid + "/fastq_screen/",
-    # container: "docker://rroutsong/dmux_ngsqc:0.0.1",
     containerized: "/data/OpenOmics/SIFs/dmux_ngsqc_0.0.1.sif"
```
Comment (skchronicles) [inline src/Dmux/workflow/ngs_qaqc/fastq.smk:44]: We need to add an option to point to a sif cache and dynamically resolve one of the following: a local SIF on the file-system or a URI to pull an image from Dockerhub.

File: bin/dmux.py
Comment (skchronicles) [inline bin/dmux.py:1]: Thank you for updating the dry-run flag/switch name!

File: src/Dmux/workflow/ngs_qaqc/fastq.smk [hunk 2]
```diff
@@ -98,7 +100,6 @@ rule kraken_annotation:
         kraken_log          = config['out_to'] + "/{project}/" + config['run_ids'] + "/{sid}/kraken/{sid}.log",
     params:
         kraken_db           = "/data/OpenOmics/references/Dmux/kraken2/k2_pluspfp_20230605"
```
Comment (skchronicles) [inline src/Dmux/workflow/ngs_qaqc/fastq.smk:102]: We need a method to dynamically resolve the reference files.

File: src/Dmux/workflow/ngs_qaqc/qc.smk [hunk 3]
```diff
@@ -57,3 +55,46 @@ rule fastqc_trimmed:
...
+        demux_dir       = config['demux_dir'],
+        output_dir      = config['out_to'] + "/" + config['projects'] + "/" + config['run_ids'] + "/multiqc/",
+        report_title    = f"Run: {config['run_ids']}, Project: {config['projects']}",
+    containerized: "/data/OpenOmics/SIFs/dmux_ngsqc_0.0.1.sif"
+    threads: 4
+    resources: mem_mb = 8096
+    log: config['out_to'] + "/.logs/" + config['projects'] + "/" + config['run_ids'] + "/multiqc/multiqc.log"
+    shell:
+        """
+        multiqc -q -ip \
```
Comment (skchronicles) [inline src/Dmux/workflow/ngs_qaqc/qc.smk:95]: At some point, we may want to point to a MutliQC config file to clean up the general statistics table, create two sections for fastqc, and create a preferred module order in the final report.

File: src/Dmux/workflow/ngs_qaqc/fastq.smk
Comment (skchronicles) [inline src/Dmux/workflow/ngs_qaqc/fastq.smk:1]: We need to change how adapter sequences are being removed. Currently, there is a bug where the barcode sequences from Illumina's sample sheet (i7/i5) sequences are being passed to fastqc and fastp. These barcode sequences should be removed after bcl2fastq step and do not represent traditional library-prep-kit-specific adapter sequences that need to removed. With that being said, let's make use of fastp's auto-detect-adapter-sequences feature to remove them. We can also make use of fastqc's internal contaminates/adapters list to identify sequencing adapters.

**Review by skchronicles — CHANGES_REQUESTED** [review]

File: src/Dmux/workflow/ngs_qaqc/fastq.smk (hunk 1 above)
Comment (rroutsong) [reply src/Dmux/workflow/ngs_qaqc/fastq.smk:44]: I have a solution to this issue in the next coming PR. I have serialized the server-centric SIF directories and dynamically adding the specific server configuration at initialization time. 

Ends up like:

`containerized: server_config["sif"] + "dmux_ngsqc_0.0.1.sif"`

SIF cache is always specified at execution time through environmental variables and subprocess.

File: src/Dmux/workflow/ngs_qaqc/fastq.smk (hunk 2 above)
Comment (rroutsong) [reply src/Dmux/workflow/ngs_qaqc/fastq.smk:102]: Also addressed this in the next PR. I just kind of saved all the server resolution methods until I moved onto bigsky.

File: src/Dmux/workflow/ngs_qaqc/qc.smk (hunk 3 above)
Comment (rroutsong) [reply src/Dmux/workflow/ngs_qaqc/qc.smk:95]: This is outlined in #15

File: src/Dmux/workflow/ngs_qaqc/fastq.smk
Comment (rroutsong) [reply src/Dmux/workflow/ngs_qaqc/fastq.smk:1]: Here's fastp rule in new branch `master_job_and_bigsky`:
```
    shell:
        """
        fastp \
        --detect_adapter_for_pe \
        --in1 {input.in_read1} --in2 {input.in_read2} \
        --out1 {output.out_read1} \
        --out2 {output.out_read2} \
        --html {output.html} \
        --json {output.json} \
        """
```

Fastqc:
```
    shell:
        """
        mkdir -p {params.output_dir}
        fastqc -o {params.output_dir} -t {threads} {input.samples}
        """
```

FastQC before trim depends on demuxed reads, after trimmed depends on trimmed reads file.

**Comment by skchronicles** [comment]
Will address some of these comments/issues in the next PR.

**Review by skchronicles — APPROVED** [review]

**Pull request merged**